### ISSUE TITLE
Questions

### ISSUE THREAD

**Issue by wcott** [sford/yet-another-commit-checker#109]
Sorry for the ambiguous title, I just have a few questions regarding several different parts of your plugins if you wouldn't mind answering them. I can't seem to label this as a "questions" bug, sorry for that as well. 

1. You seem to create a lot of classes/methods that atlassian provides through their API. While not being entirely the same, they are very similar. The CommitService in 3.11.1 has several methods like "getCommitsBetween" that allow you to grab commits from a refChange object. Sorry I can't give specific links, atlassian has done something strange with their API reference since I developed my plugin when they made the change to Bitbucket. I'm finding this is the case with a lot of commit checker plugins for Stash/Bitbucket. People are creating their own classes/methods for munging data from the commits whereas when I made my plugin entirely with classes/methods provided from atlassian (probably just a novice mistake). Is there a specific reason you did this? 

2. What's the purpose of the config validator class? did you encounter some reason/error in your testing for it? 

3. My plugin does a lot of event handling. I see in the BranchCreationListener you register an event handler and then turn it into a hook? What's going on in [here](https://github.com/sford/yet-another-commit-checker/blob/master/src/main/java/com/isroot/stash/plugin/YaccBranchCreationListener.java#L41-L78)

Thanks for taking the time to answer my questions.

**Comment by sford** [comment]
Hi @wcott,

Thanks for the questions!

1. Regarding `CommitService` and `getCommitsBetween` usage, YACC does indeed depend on this API. YACC `CommitsService` (with an _s_ in Commits... probably poorly named :-) class is just essentially a wrapper around `getCommitsBetween` that does some extra YACC-specific stuff before and after calling `getCommitsBetween`. This helps clean up some other code because this extra stuff is abstracted away from the calling method.

2. The config validator class validates the input fields on the settings page. This is important because without, it is possible that bunk data (like invalid regex's) could cause YACC to throw exceptions when checking commits.

3. Great question. YACC can be configured in 2 ways... you can either configure it per-repository (repository hook settings), or you can configure YACC globally (Manage add-on's -> settings). If you have a lot of repositories, global configuration allows you to configure YACC once and have it apply everywhere instead of having to configure each repository individually. If YACC is configured for a repository, than the repository settings trump the global settings. This is what the `BranchCreationListener` code is doing... it first tries to pull the repository settings to see if they exist. If not, it defaults to the global settings.

Hope this is helpful! Cheers, 🍺

**Comment by wcott** [comment]
Got one more question:

Your plugin description states that you only ever analyze new commits which is great because each time you make a new branch Stash thinks every commit on that branch is new. I was trying to figure out how you do this, and I figured it's when you exclude all the branch names from the getsCommitsBetweenRequest with this method: https://github.com/sford/yet-another-commit-checker/blob/master/src/main/java/com/isroot/stash/plugin/CommitsServiceImpl.java#L122-L138

The only question I have about this method is that you return all the branch names, including the one the commit is one. I figured that that branch name would equate to the latest ref on that branch and I tested it too. When I construct a getCommitsBetweenBranchRequest similar to the way you do it here, I get no commits cause I'm effectively excluding and including the same commit id (I think?). Where is the magic happening? 

Thanks! 
Scott
